servant-server/src/Servant/Server/Internal/RoutingApplication.hs

```diff
@@ -30,39 +26,12 @@ type RoutingApplication =
   -> (RouteResult Response -> IO ResponseReceived) -> IO ResponseReceived
 
 -- | A wrapper around @'Either' 'RouteMismatch' a@.
-newtype RouteResult a =
-  RR { routeResult :: Either RouteMismatch a }
-  deriving (Eq, Show, Functor, Applicative)
-
--- | If we get a `Right`, it has precedence over everything else.
---
--- This in particular means that if we could get several 'Right's,
--- only the first we encounter would be taken into account.
-instance Monoid (RouteResult a) where
-  mempty = RR $ Left mempty
-
-  RR (Left x)  `mappend` RR (Left y)  = RR $ Left (x <> y)
-  RR (Left _)  `mappend` RR (Right y) = RR $ Right y
-  r            `mappend` _            = r
-
--- Note that the ordering of the constructors has great significance! It
--- determines the Ord instance and, consequently, the monoid instance.
-data RouteMismatch =
-    NotFound           -- ^ the usual "not found" error
-  | WrongMethod        -- ^ a more informative "you just got the HTTP method wrong" error
-  | UnsupportedMediaType -- ^ request body has unsupported media type
-  | InvalidBody String -- ^ an even more informative "your json request body wasn't valid" error
-  | HttpError Status (Maybe BL.ByteString)  -- ^ an even even more informative arbitrary HTTP response code error.
-  deriving (Eq, Ord, Show)
-
-instance Monoid RouteMismatch where
-  mempty = NotFound
-  -- The following isn't great, since it picks @InvalidBody@ based on
-  -- alphabetical ordering, but any choice would be arbitrary.
-  --
-  -- "As one judge said to the other, 'Be just and if you can't be just, be
-  -- arbitrary'" -- William Burroughs
-  mappend = max
+data RouteResult a =
+    Retriable ServantErr      -- ^ Keep trying other paths. The @ServantErr@
+                              -- should only be 404 or 405.
+  | NonRetriable ServantErr   -- ^ Stop trying.
+  | HandlerVal a
+  deriving (Eq, Show, Read, Functor)
```

Comment:
This brilliantly cleans up the internals of servant quite significantly. However, I suggest we rename the constructors of this data type. `Retriable`/`NonRetriable` hold `ServantErr`s, and the errors themselves are not "retriable" or "non retriable". They are either "final" (as in "this is the definitive error we will return") or not. I suggest we directly reflect this in the names.

How about:

``` haskell
data RouteResult a
  = Fail ServantErr
  | Fatal ServantErr
  | Routed a
```

PS: I've also been wondering whether we really want `ServantErr` instead of `Response` so that folks could just use the functions from WAI and not record-update our `errXXX`.